src/ui/profile/AddProfile.js

```diff
@@ -0,0 +1,146 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import styles from '../styles/profile.module.css';
+import ProfileModel from '../../models/profile';
+import type from '../../utils/types';
+
+const BEYOND_ID_LENGTH = 8;
+
+function AddProfileForms(props) {
+  const { onAddProfile, onCancel, onError } = props;
+  const [charId, setCharId] = useState('');
+
+  const validateInputId = (input) => {
+    const parsedId = parseInt(input, 10);
+    if (input.length !== BEYOND_ID_LENGTH || !type.isNumber(parsedId)) {
+      throw new Error('Invalid DnD Beyond Character ID.');
+    }
+    return parsedId;
+  };
+
+  const handleAddCharacter = () => {
+    let id;
+    try {
+      id = validateInputId(charId);
+    } catch (err) {
+      onError(err.message);
+      return;
+    }
+
+    const url = `https://www.dndbeyond.com/character/${id}/json`;
+
+    try {
+      onAddProfile(id, fetch(url)
+        .then((response) => response.json())
+        .catch(() => Promise.reject(new Error('Profile not found.')))
+        .then((jsonData) => {
+          if (jsonData.errorCode === 404) {
+            const strs = [
+              'Could not retrieve profile.',
+              'Please verify your character privacy is set to public.',
+            ];
+            throw new Error(strs.join('\n'));
+          }
+          const avatar = jsonData.avatarUrl || '';
+          const level = jsonData.classes.reduce((total, classObj) => {
+            const classLevel = parseInt(classObj.level, 10);
+            return total + classLevel;
+          }, 0);
+          return ProfileModel({
+            id,
+            name: jsonData.name,
+            avatar,
+            level,
+          });
+        }));
+    } catch (err) {
+      if (err.message === 'ALREADY_EXISTS') {
+        onError('Character already added.');
+      } else {
+        throw err;
+      }
+    }
+  };
+
+  return (
+    <div>
+      <form className={styles.InputForm}>
```

Comment:
This doesn't need to be a `<form />`. It can just be a `<div />`. Forms are generally only useful if you are actually submitting the form data to a server.

History lesson!

Forms come from the early HTML days before you could just send HTTP requests from JS. They were basically the only way for developers to allow user input to be submitted to the backend because you encoded the HTTP method and path into the form, and when the user clicked the submit button the browser would serialize the form field values and send them to the server for you. The response of that request would then replace the current page. This old model is called a "round-trip" web application.

Then XHR (and now `fetch()`) became a thing so now JS can send asynchronous requests without having to reload the page. People took more and more advantage of this and realized that they could bundle their entire application into one HTML page even if their site has multiple routes. This new model is called a "single-page" web application (SPA), and is generally the only way people do things nowadays.

Reply:
Yeah it looks like the `form` is not doing anything for me in this case. Is it generally better to default to `div`s since they don't have unnecessary functionality?

Reply:
There are two element types that are generally considered to be "default" when you are looking for a container and nothing more:
* `<div />` is the default for most situations, particularly where you are looking for `display: block` (elements that are arranged vertically)
* `<span />` is the default when you are looking for `display: inline` (elements that are arranged horizontally inline with each other, wrapping to the next line if there is room)